configs/example/gem5_library/x86-global-inst-tracker.py

```diff
@@ -0,0 +1,159 @@
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+
+This script demonstrates how to use global and local instruction trackers to
+monitor and control the simulation based on the number of instructions
+committed.
+
+This script will create a global instruction tracker to manage each local
+instruction tracker and connects each local instruction tracker to a core.
+The global instruction tracker will raise an event when the number of
+instructions committed across all cores reaches a certain threshold.
+
+In this script, we expect to start monitorning the instruction committed from
+the start of the simulation and raise a SIMPOINT_BEGIN exit event when all
+cores in combination have committed 100,000,000 instructions. Then, we will
+change the threshold to 20,000 instructions and raise another SIMPOINT_BEGIN
+event when the new threshold is reached. Finally, we will stop listening to
+instructions.
+
+Usage:
+------
+
+scons build/X86/gem5.opt
+./build/X86/gem5.opt [--debug-flags=InstTracker] \
+    configs/example/gem5_library/x86-global-inst-tracker.py
+
+"""
+
+import m5
+from m5.objects import (
+    GlobalInstTracker,
+    LocalInstTracker,
+)
+
+from gem5.components.boards.simple_board import SimpleBoard
+from gem5.components.cachehierarchies.classic.private_l1_cache_hierarchy import (
+    PrivateL1CacheHierarchy,
+)
+from gem5.components.memory.single_channel import SingleChannelDDR4_2400
+from gem5.components.processors.cpu_types import CPUTypes
+from gem5.components.processors.simple_processor import SimpleProcessor
+from gem5.isas import ISA
+from gem5.resources.resource import obtain_resource
+from gem5.simulate.exit_event import ExitEvent
+from gem5.simulate.simulator import Simulator
+
+cache_hierarchy = PrivateL1CacheHierarchy(
+    l1d_size="64kB",
+    l1i_size="64kB",
+)
+
+memory = SingleChannelDDR4_2400("1GB")
+
+processor = SimpleProcessor(cpu_type=CPUTypes.TIMING, num_cores=9, isa=ISA.X86)
+
+# setup instruction tracker
+
+# first, we need to create a global instruction tracker
+global_inst_tracker = GlobalInstTracker(
+    # threshold to trigger the event
+    inst_threshold=100_000_000
+)
+
+# then, we create a local instruction tracker for each core
+# and store them in a list
+all_trackers = []
+
+for core in processor.get_cores():
+    tracker = LocalInstTracker(
+        # we pass in the glboal instruction tracker to the local one
+        global_inst_tracker=global_inst_tracker,
+        # this parameter tells the tracker to start listening to instructions
+        # from the beginning of the simulation. If set to False, the tracker
+        # will
+        start_listening=True,
+    )
+    # we attach the tracker to the core
+    core.core.probeListener = tracker
+    # then store the tracker in the list
+    all_trackers.append(tracker)
```

Comment:
For one core, it has an existing function to check the instruction counts (https://github.com/gem5/gem5/blob/stable/src/cpu/BaseCPU.py#L73) and a way to set an exit event at instructions (https://github.com/gem5/gem5/blob/stable/src/cpu/BaseCPU.py#L72). 
However, there isn't a way to schedule an instruction exit event for multiple cores. This PR is to provide a way for the multiple cores' scenario. We can have `set_exit_at_inst` method exits in both core level and processor level if they call different functions to schedule the exit event. 
I don't have a strong opinion on this.

Reply:
If it's all the same 1 would prefer my suggested approach

Reply:
I think it can add significant overhead. In general, it adds more executions to every instruction committed. I believe there is a balanced solution for this interface, but we should move this discussion to the future PR for its stdlib interface.